scripts/zones/Batallia_Downs_[S]/npcs/Thorben.lua

```diff
@@ -0,0 +1,65 @@
+-----------------------------------
+-- Area: Batallia Downs (S)
+--  NPC: Thorben
+-- Type: Quest NPC
+-- !pos  175.346,8.038,-419.244 84
+-----------------------------------
+require("scripts/globals/keyitems")
+require("scripts/globals/settings")
+require("scripts/globals/quests")
+require("scripts/globals/npc_util")
+-----------------------------------
+
+function onTrade(player,npc,trade)
+end
+
+function onTrigger(player,npc)
+    local lostInTranslocation = player:getQuestStatus(CRYSTAL_WAR, dsp.quest.id.crystalWar.LOST_IN_TRANSLOCATION)
+    local leftMapPiece = player:hasKeyItem(dsp.ki.LEFT_MAP_PIECE)
+    local middleMapPiece = player:hasKeyItem(dsp.ki.MIDDLE_MAP_PIECE)
+    local rightMapPiece = player:hasKeyItem(dsp.ki.RIGHT_MAP_PIECE)
+
+    if lostInTranslocation == QUEST_COMPLETED then
+        player:startEvent(108)
+    elseif lostInTranslocation == QUEST_ACCEPTED and leftMapPiece and middleMapPiece and rightMapPiece then
+        player:startEvent(107)
+    elseif player:getVar("lostInTranslocationCS") == 1 and (leftMapPiece or middleMapPiece or rightMapPiece) then
+        player:startEvent(106)
+    elseif lostInTranslocation == QUEST_ACCEPTED and (leftMapPiece or middleMapPiece or rightMapPiece) then
+        player:startEvent(105)
+    elseif lostInTranslocation == QUEST_ACCEPTED and not leftMapPiece and not middleMapPiece and not rightMapPiece then
+        player:startEvent(104)
+    else
+        player:startEvent(103)
+    end
+end
+
+function onEventUpdate(player,csid,option)
+end
+
+function onEventFinish(player,csid,option)
+    if csid == 107 then
+        if player:hasKeyItem(dsp.ki.MAP_OF_GRAUBERG) then
+            npcUtil.completeQuest(player, CRYSTAL_WAR, dsp.quest.id.crystalWar.LOST_IN_TRANSLOCATION, {
+                gil = 2000,
+                xp = 2000,
+                var = "lostInTranslocationCS"
+            })
+            player:delKeyItem(dsp.ki.LEFT_MAP_PIECE)
+            player:delKeyItem(dsp.ki.MIDDLE_MAP_PIECE)
+            player:delKeyItem(dsp.ki.RIGHT_MAP_PIECE)
+        else
+            npcUtil.completeQuest(player, CRYSTAL_WAR, dsp.quest.id.crystalWar.LOST_IN_TRANSLOCATION, {
+                keyItem = dsp.ki.MAP_OF_GRAUBERG,
+                var = "lostInTranslocationCS"
+            })
+            player:delKeyItem(dsp.ki.LEFT_MAP_PIECE)
+            player:delKeyItem(dsp.ki.MIDDLE_MAP_PIECE)
+            player:delKeyItem(dsp.ki.RIGHT_MAP_PIECE)
+        end
+    elseif csid == 105 then
+        player:setVar("lostInTranslocationCS", 1)
```

Comment:
I'm curious, what's this var being used for? What's the difference between event 105 and 106? (There apparently isn't a dump for Batalia [S]).

Reply:
@ibm2431  the var is used so if you talk to the NPC at csid == 105 it will give you a longer CS, compared if you talk to him again, he will still refer to the same dialog that was on csid == 105, but the dialog will only be a 1- 2 sentence dialog.

Reply:
I suppose the var is fair then; just making sure it's required!

(Although, does it only play just that one time, or one per each zone? Would a entity local var be preferable?)

Reply:
@ibm2431  it doesn't play only once as NPC doesn't have a "standard greeting", instead he is stuck on that until the keyitems attained changed to have a few, or all of them. So for instance if the PC would leave to do something else like other quests or missions, and then eventually come back to this quest it should still play csid == 106.

With my explanation, do you think entity local var is still needed? Is there an example of a entity local var in one of the files? (sorry for asking but usually i refer to other quests people have done, to guide myself) so that way if I do another quest I can think of using it if needed 😸️

Reply:
If the NPC should always say that even after the player zones and comes back, then I guess it should be a DB var.

But if you're interested in entity local vars (which DSP currently refers to as "local vars"; terminology which can be confusing), I think the [most recent examples of `setLocalVar` and `getLocalVar` can be found with the Moghouse quests](https://github.com/DarkstarProject/darkstar/blob/master/scripts/globals/moghouse.lua).

You use them just like normal vars that'd otherwise go in the SQL DB, but the difference is that these vars are stored in memory (for each specific entity), and are lost upon zoning. They can also be used for mobs as well as players, which may be useful if the mob needd to do anything out of the ordinary.

Reply:
Awesome info 😸️ Thank you @ibm2431  I'll keep it in mind next time if it doesnt need to be a DB var.